### DISCUSSION TITLE
Client Extension Types - different Prisma versions

### DISCUSSION THREAD

**Discussion by dennemark** [prisma/prisma#24987] [Q&A] [answered]
### Question

Hi,
I am developing a prisma client extension  https://github.com/dennemark/prisma-extension-casl
I have a client function that returns the extended client like `client.$casl((a)=>a).user.findMany()`. When I use my extension in a project, the types are missing though, after calling the $casl function. So no user will be shown.

The reason for this are different Prisma versions. The old version (5.14.0) uses `DynamicClientExtensionThis` with 3 arguments and the new version (5.17.0) uses 4. Now I am wondering, how can have backward or forward compatible types?
 
 In my case I was so egoistic to downgrade the dependency of my extension to my prisma version. Not very helpful for other users :(

How would you do it? Release another version of the extension with the newer prisma client? What happens if a bug appear.

### How to reproduce (optional)

`DynamicClientExtensionThis` with 3 arguments in version 5.14.0 vs 4 in version 5.17.0

### Information about Prisma Schema, Client Queries and Environment (optional)

5.14.0 vs 5.17.0

**Comment by ludralph** [comment, marked as answer]
Hi @dennemark 

One approach is to release different versions of your extension that are compatible with specific versions of Prisma. This way, users can choose the version of your extension that matches their Prisma version.
For example:
```
prisma-extension-casl@1.x for Prisma 5.14.0
prisma-extension-casl@2.x for Prisma 5.17.0
```
This approach ensures that users can always find a compatible version of your extension, but it requires maintaining multiple versions.
You would need to clearly document the compatibility of your extension with different Prisma versions. This includes specifying which version of your extension works with which version of Prisma. 

Another approach is to use TypeScript's conditional types to handle the differences in the `DynamicClientExtensionThis` type between Prisma versions. This can be more complex but allows you to maintain a single version of your extension that works across multiple Prisma versions.

Here's a simplified example of how you might handle different versions:
```ts
// Define a type that checks the Prisma version and adjusts accordingly
type PrismaClientExtensionThis<T> = T extends { version: '5.14.0' }
  ? DynamicClientExtensionThis<Arg1, Arg2, Arg3>
  : DynamicClientExtensionThis<Arg1, Arg2, Arg3, Arg4>;

// Use the type in your extension
function $casl<T>(client: PrismaClientExtensionThis<T>) {
  // Your implementation here
}
```
The best approach depends on your specific needs and the needs of your users.

**Reply by dennemark** [reply]
Thanks so mich for your reply! I might go for the versioning approach since it seems to be a breaking change. Though I also might try the typescript approach :)
